c++/src/kj/compat/brotli.c++

```diff
@@ -0,0 +1,355 @@
+    KJ_ASSERT(BrotliEncoderSetParameter(cctx, BROTLI_PARAM_LGWIN, windowBits) == BROTLI_TRUE);
+  } else {
+    compressing = false;
+    // In the decoder, we manually check that the stream does not have a higher window size than
+    // requested and reject it otherwise, no way to automate this step.
+    // By default, we accept streams with a window size up to (1 << KJ_BROTLI_MAX_DEC_WBITS),
+    // this is more than the default window size for compression (i.e. KJ_BROTLI_DEFAULT_WBITS)
+    windowBits = _windowBits.orDefault(_::KJ_BROTLI_MAX_DEC_WBITS);
+    KJ_REQUIRE(windowBits >= BROTLI_MIN_WINDOW_BITS && windowBits <= BROTLI_MAX_WINDOW_BITS,
+        "invalid brotli window size (window bits", windowBits, ")");
+    dctx = BrotliDecoderCreateInstance(nullptr, nullptr, nullptr);
+    KJ_REQUIRE(dctx, "brotli state allocation failed");
+  }
+}
+
+BrotliOutputContext::~BrotliOutputContext() noexcept(false) {
+  compressing ? BrotliEncoderDestroyInstance(cctx) : BrotliDecoderDestroyInstance(dctx);
+}
+
+void BrotliOutputContext::setInput(const void* in, size_t size) {
+  next_in = reinterpret_cast<const byte*>(in);
+  available_in = size;
+}
+
+kj::Tuple<bool, kj::ArrayPtr<const byte>> BrotliOutputContext::pumpOnce(BrotliEncoderOperation flush) {
+  byte* next_out = buffer;
+  size_t available_out = sizeof(buffer);
+  // Brotli does not accept a null input pointer; make sure there is a valid pointer even if we are
+  // not actually reading from it.
+  if (!next_in) {
+    KJ_ASSERT(available_in == 0);
+    next_in = buffer;
+  }
+
+  if (!compressing) {
+    // Check window bits
+    if (firstInput) {
+      firstInput = false;
+      int streamWbits = get_brotli_window_bits(next_in[0]);
+      KJ_REQUIRE(streamWbits <= windowBits,
+        "brotli window size too big", (1 << streamWbits));
+    }
+    BrotliDecoderResult result =
+        BrotliDecoderDecompressStream(dctx, &available_in, &next_in, &available_out, &next_out,
+                                      nullptr);
+    if (result == BROTLI_DECODER_RESULT_ERROR) {
+        // Note: Unlike BrotliInputStream, this will implicitly reject trailing data during
+        // decompression, matching the behavior for gzip.
+        KJ_FAIL_REQUIRE("brotli decompression failed",
+                        BrotliDecoderErrorString(BrotliDecoderGetErrorCode(dctx)));
+    }
+  } else {
+    BROTLI_BOOL result = BrotliEncoderCompressStream(cctx, flush, &available_in, &next_in,
+                                                     &available_out, &next_out, nullptr);
+    KJ_REQUIRE(result == BROTLI_TRUE, "brotli compression failed");
+  }
+
+  // The 'ok' parameter represented by the first parameter of the tuple indicates that pumpOnce()
+  // should be called again as more output data can be produced. This is the case when the stream
+  // is not finished and there is either pending output data (that didn't fit into the buffer) or
+  // input that has not been processed yet.
```

Comment:
We tend not to use tuples in return types, because they're not very self-documenting. Instead, I'd probably express the return type as something like this, integrating your comment here.
```cpp
struct PumpOnceResult {
  bool callAgain;
  // If true, `pumpOnce()` should be called again, as more output data can be produced. This is the case when ...

  kj::ArrayPtr<const byte> output;
  // A view into BrotliOutputContext's internal buffer containing the encoded/decoded output.
};
```

I see that the tuple pattern was copied from the gzip implementation, so I'll leave it up to you if you want to change this, change both, or leave as-is.